### ISSUE TITLE
Behave tests should implement a definitive interlock with Docker-compose

### ISSUE THREAD

**Issue by bcbrock** [hyperledger-archives/fabric#1234]
I have been having many problems today with trying to get the **behave** tests to run. The problem is that the assumptions on how long it is reasonable to wait between 1) the Docker-compose call terminating, and 2) the peers and membersrvc servers in all containers actually being up and ready to accept transactions, is too short. 

Time-based assumptions like this are not robust in general. For this particular problem, I believe that there are several techniques that  could be used to guarantee that a Docker-compose network was actually fully up and ready to accept transactions before allowing the test to continue.

I don't have time to work on this however. Instead, I am going to submit a PR that centralizes the decision  on when it is safe to proceed after starting a Docker-compose network. For now, this is is simply a longer timeout. Hopefully in the future someone will develop a more robust and definitive solution.

**Comment by bcbrock** [comment]
I apologize; I overlooked an important piece of the puzzle yesterday. One major reason that the current timeouts no longer work for me is that in networks the include **membersrvc**, I need to hold off the startup of the peers for a few seconds to give the **membersrvc** server time to come up. Otherwise I get intermittent "no route to host" errors from the peers. Again, there must be better solutions to this issue than a simple timeout, but in the meantime we need the tests to be more reliable.

**Comment by binhn** [comment]
@jeffgarratt is there a better way to detect when a container is up?  perhaps we should send membersrvc a ping to ensure that it is on?

**Comment by bcbrock** [comment]
We now have the `busywork/bin/wait-for-it` script, that does what @binhn suggested. The networks that busywork creates use `wait-for-it` to implement the definitive interlock.

**Comment by christo4ferris** [comment]
@bcbrock I tend to agree that timing based testing can be non-deterministic if there are other variables such as cpu, disk or network load from noisy neighbors. However, I think that as you have busywork under control, now, whether we might close this until we perceive the problem to be something that demands mitigation?

**Comment by bcbrock** [comment]
I would not close this issue - This is a real problem with known, workable solutions that will improve the robustness and reliability of the tests. Better would be to put a "help wanted" on it.
